library/src/scripts/components/siteNav/SiteNav.tsx

```diff
@@ -0,0 +1,146 @@
+/*
+ * @author Stéphane LaFlèche <stephane.l@vanillaforums.com>
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license GPL-2.0-only
+ */
+
+import * as React from "react";
+import { t } from "../../application";
+import classNames from "classnames";
+import { getRequiredID } from "@library/componentIDs";
+import { withRouter, RouteComponentProps } from "react-router-dom";
+import SiteNavNode from "@library/components/siteNav/SiteNavNode";
+import TabHandler from "@library/TabHandler";
+
+interface IProps extends RouteComponentProps<{}> {
+    name: string;
+    className?: string;
+    children: any[];
+}
+
+export interface IState {
+    id: string;
+}
+
+/**
+ * Recursive component to generate site nav
+ * No need to set "counter". It will be set automatically. Kept optional to not need to call it on the top level. Used for React's "key" values
+ */
+export class SiteNav extends React.Component<IProps, IState> {
+    public constructor(props) {
+        super(props);
+        this.state = {
+            id: getRequiredID(props, "siteNav"),
```

Comment:
Why our we using this instead of `uniqueIDFromPrefix()` as a property initializer? Since none if this actually comes from props it could doesn't need `getRequiredId`. In fact if you added the type `IProps` on your constructor this would fail the type checker because there is no ID there.

We could replace this constructor override with

```ts
class SiteNav extends React.Component {
    private id = uniqueIDFromPrefix("siteNav");
}
```

and access it with `this.id`.